bytecomp/matching.ml

```diff
@@ -1562,59 +1672,75 @@ let inline_lazy_force arg loc =
          tables (~ 250 elts); conditionals are better *)
       inline_lazy_force_cond arg loc
 
-let make_lazy_matching def = function
+let make_lazy_matching loc def = function
     [] -> fatal_error "Matching.make_lazy_matching"
-  | (arg,_mut) :: argl ->
+  | (arg, _mut, _arg_loc) :: argl ->
       { cases = [];
         args =
-          (inline_lazy_force arg Location.none, Strict) :: argl;
+          (inline_lazy_force arg loc, Strict, loc) :: argl;
         default = make_default matcher_lazy def }
 
 let divide_lazy p ctx pm =
   divide_line
     (filter_ctx p)
-    make_lazy_matching
+    (make_lazy_matching p.pat_loc)
     get_arg_lazy
     p ctx pm
 
 (* Matching against a tuple pattern *)
 
 
 let get_args_tuple arity p rem = match p with
-| {pat_desc = Tpat_any} -> omegas arity @ rem
+| {pat_desc = Tpat_any} -> omegas p.pat_loc arity @ rem
 | {pat_desc = Tpat_tuple args} ->
     args @ rem
 | _ ->  assert false
 
 let matcher_tuple arity p rem = match p.pat_desc with
 | Tpat_or (_,_,_)     -> raise OrPat
 | Tpat_any
-| Tpat_var _ -> omegas arity @ rem
+| Tpat_var _ -> omegas p.pat_loc arity @ rem
 | Tpat_tuple args when List.length args = arity -> args @ rem
 | _ ->  raise NoMatch
 
-let make_tuple_matching loc arity def = function
+let make_tuple_matching arity field_pat_locs def = function
     [] -> fatal_error "Matching.make_tuple_matching"
-  | (arg, _mut) :: argl ->
+  | (arg, _mut, _arg_loc) :: argl ->
       let rec make_args pos =
         if pos >= arity
         then argl
-        else (Lprim(Pfield pos, [arg], loc), Alias) :: make_args (pos + 1) in
+        else begin
+          assert (pos >= 0 && pos < Array.length field_pat_locs);
+          let arg_loc = field_pat_locs.(pos) in
+          (Lprim(Pfield pos, [arg], arg_loc), Alias, arg_loc)
+            :: make_args (pos + 1)
+        end
+      in
       {cases = []; args = make_args 0 ;
         default=make_default (matcher_tuple arity) def}
 
 
 let divide_tuple arity p ctx pm =
+  let field_pat_locs =
+    match p.pat_desc with
+    | Tpat_tuple args -> Array.of_list (List.map (fun pat -> pat.pat_loc) args)
+    | _ -> Misc.fatal_error "Only [Tpat_any] or [Tpat_tuple] expected"
+  in
```

Comment:
Defining and passing `field_pats` is fine, but note that the code would be more concise and arguably more future-proof if you passed `args` directly. (Also on the naming front, `fields` would be just a good a name as `field_pats`: we are decomposing patterns so of course constructor fields are going to be patterns as well.)

I don't have a strong opinion on this one, but I would recommend trying to just pass `args` (renamed into `fields`) directly. I think the code would more concise and thus clearer.

Reply:
Not sure this is worth doing, looking at the call site.  Also, if you pass that variable separately, it raises the possibility of it being different from what is inside `p`, whereas that cannot happen at present.